// antd/es/transfer/list.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _createSuper from "@babel/runtime/helpers/esm/createSuper";
import DownOutlined from "@ant-design/icons/es/icons/DownOutlined";
import classNames from 'classnames';
import omit from "rc-util/es/omit";
import * as React from 'react';
import Checkbox from '../checkbox';
import Dropdown from '../dropdown';
import { isValidElement } from '../_util/reactNode';
import { groupKeysMap } from '../_util/transKeys';
import DefaultListBody, { OmitProps } from './ListBody';
import Search from './search';
const defaultRender = () => null;
function isRenderResultPlainObject(result) {
  return !!(result && !isValidElement(result) && Object.prototype.toString.call(result) === '[object Object]');
}
function getEnabledItemKeys(items) {
  return items.filter(data => !data.disabled).map(data => data.key);
}
let TransferList = /*#__PURE__*/function (_React$PureComponent) {
  _inherits(TransferList, _React$PureComponent);
  var _super = _createSuper(TransferList);
  function TransferList(props) {
    var _this;
    _classCallCheck(this, TransferList);
    _this = _super.call(this, props);
    _this.defaultListBodyRef = /*#__PURE__*/React.createRef();
    // =============================== Filter ===============================
    _this.handleFilter = e => {
      const {
        handleFilter
      } = _this.props;
      const {
        target: {
          value: filterValue
        }
      } = e;
      _this.setState({
        filterValue
      });
      handleFilter(e);
    };
    _this.handleClear = () => {
      const {
        handleClear
      } = _this.props;
      _this.setState({
        filterValue: ''
      });
      handleClear();
    };
    _this.matchFilter = (text, item) => {
      const {
        filterValue
      } = _this.state;
      const {
        filterOption
      } = _this.props;
      if (filterOption) {
        return filterOption(filterValue, item);
      }
      return text.includes(filterValue);
    };
    // =============================== Render ===============================
    _this.renderListBody = (renderList, props) => {
      let bodyContent = renderList ? renderList(props) : null;
      const customize = !!bodyContent;
      if (!customize) {
        bodyContent = /*#__PURE__*/React.createElement(DefaultListBody, Object.assign({
          ref: _this.defaultListBodyRef
        }, props));
      }
      return {
        customize,
        bodyContent
      };
    };
    _this.renderItem = item => {
      const {
        render = defaultRender
      } = _this.props;
      const renderResult = render(item);
      const isRenderResultPlain = isRenderResultPlainObject(renderResult);
      return {
        renderedText: isRenderResultPlain ? renderResult.value : renderResult,
        renderedEl: isRenderResultPlain ? renderResult.label : renderResult,
        item
      };
    };
    _this.getSelectAllLabel = (selectedCount, totalCount) => {
      const {
        itemsUnit,
        itemUnit,
        selectAllLabel
      } = _this.props;
      if (selectAllLabel) {
        return typeof selectAllLabel === 'function' ? selectAllLabel({
          selectedCount,
          totalCount
        }) : selectAllLabel;
      }
      const unit = totalCount > 1 ? itemsUnit : itemUnit;
      return /*#__PURE__*/React.createElement(React.Fragment, null, (selectedCount > 0 ? `${selectedCount}/` : '') + totalCount, " ", unit);
    };
    _this.state = {
      filterValue: ''
    };
    return _this;
  }
  _createClass(TransferList, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this.triggerScrollTimer);
    }
  }, {
    key: "getCheckStatus",
    value: function getCheckStatus(filteredItems) {
      const {
        checkedKeys
      } = this.props;
      if (checkedKeys.length === 0) {
        return 'none';
      }
      const checkedKeysMap = groupKeysMap(checkedKeys);
      if (filteredItems.every(item => checkedKeysMap.has(item.key) || !!item.disabled)) {
        return 'all';
      }
      return 'part';
    }
    // ================================ Item ================================
  }, {
    key: "getFilteredItems",
    value: function getFilteredItems(dataSource, filterValue) {
      const filteredItems = [];
      const filteredRenderItems = [];
      dataSource.forEach(item => {
        const renderedItem = this.renderItem(item);
        const {
          renderedText
        } = renderedItem;
        // Filter skip
        if (filterValue && !this.matchFilter(renderedText, item)) {
          return null;
        }
        filteredItems.push(item);
        filteredRenderItems.push(renderedItem);
      });
      return {
        filteredItems,
        filteredRenderItems
      };
    }
  }, {
    key: "getListBody",
    value: function getListBody(prefixCls, searchPlaceholder, filterValue, filteredItems, notFoundContent, filteredRenderItems, checkedKeys, renderList, showSearch, disabled) {
      const search = showSearch ? /*#__PURE__*/React.createElement("div", {
        className: `${prefixCls}-body-search-wrapper`
      }, /*#__PURE__*/React.createElement(Search, {
        prefixCls: `${prefixCls}-search`,
        onChange: this.handleFilter,
        handleClear: this.handleClear,
        placeholder: searchPlaceholder,
        value: filterValue,
        disabled: disabled
      })) : null;
      const {
        bodyContent,
        customize
      } = this.renderListBody(renderList, Object.assign(Object.assign({}, omit(this.props, OmitProps)), {
        filteredItems,
        filteredRenderItems,
        selectedKeys: checkedKeys
      }));
      const getNotFoundContent = () => {
        const contentIndex = this.props.direction === 'left' ? 0 : 1;
        return Array.isArray(notFoundContent) ? notFoundContent[contentIndex] : notFoundContent;
      };
      let bodyNode;
      // We should wrap customize list body in a classNamed div to use flex layout.
      if (customize) {
        bodyNode = /*#__PURE__*/React.createElement("div", {
          className: `${prefixCls}-body-customize-wrapper`
        }, bodyContent);
      } else {
        bodyNode = filteredItems.length ? bodyContent : /*#__PURE__*/React.createElement("div", {
          className: `${prefixCls}-body-not-found`
        }, getNotFoundContent());
      }
      return /*#__PURE__*/React.createElement("div", {
        className: classNames(showSearch ? `${prefixCls}-body ${prefixCls}-body-with-search` : `${prefixCls}-body`)
      }, search, bodyNode);
    }
  }, {
    key: "getCheckBox",
    value: function getCheckBox(_ref) {
      let {
        filteredItems,
        onItemSelectAll,
        disabled,
        prefixCls
      } = _ref;
      const checkStatus = this.getCheckStatus(filteredItems);
      const checkedAll = checkStatus === 'all';
      const checkAllCheckbox = /*#__PURE__*/React.createElement(Checkbox, {
        disabled: disabled,
        checked: checkedAll,
        indeterminate: checkStatus === 'part',
        className: `${prefixCls}-checkbox`,
        onChange: () => {
          // Only select enabled items
          onItemSelectAll(filteredItems.filter(item => !item.disabled).map(_ref2 => {
            let {
              key
            } = _ref2;
            return key;
          }), !checkedAll);
        }
      });
      return checkAllCheckbox;
    }
  }, {
    key: "render",
    value: function render() {
      const {
        filterValue
      } = this.state;
      const {
        prefixCls,
        dataSource = [],
        titleText = '',
        checkedKeys,
        disabled,
        footer,
        showSearch = false,
        style,
        searchPlaceholder,
        notFoundContent,
        selectAll,
        selectCurrent,
        selectInvert,
        removeAll,
        removeCurrent,
        renderList,
        onItemSelectAll,
        onItemRemove,
        showSelectAll = true,
        showRemove,
        pagination,
        direction
      } = this.props;
      // Custom Layout
      const footerDom = footer && (footer.length < 2 ? footer(this.props) : footer(this.props, {
        direction
      }));
      const listCls = classNames(prefixCls, {
        [`${prefixCls}-with-pagination`]: !!pagination,
        [`${prefixCls}-with-footer`]: !!footerDom
      });
      // ====================== Get filtered, checked item list ======================
      const {
        filteredItems,
        filteredRenderItems
      } = this.getFilteredItems(dataSource, filterValue);
      // ================================= List Body =================================
      const listBody = this.getListBody(prefixCls, searchPlaceholder, filterValue, filteredItems, notFoundContent, filteredRenderItems, checkedKeys, renderList, showSearch, disabled);
      // ================================ List Footer ================================
      const listFooter = footerDom ? /*#__PURE__*/React.createElement("div", {
        className: `${prefixCls}-footer`
      }, footerDom) : null;
      const checkAllCheckbox = !showRemove && !pagination && this.getCheckBox({
        filteredItems,
        onItemSelectAll,
        disabled,
        prefixCls
      });
      let items;
      if (showRemove) {
        items = [/* Remove Current Page */
        pagination ? {
          key: 'removeCurrent',
          onClick: () => {
            var _a;
            const pageKeys = getEnabledItemKeys((((_a = this.defaultListBodyRef.current) === null || _a === void 0 ? void 0 : _a.getItems()) || []).map(entity => entity.item));
            onItemRemove === null || onItemRemove === void 0 ? void 0 : onItemRemove(pageKeys);
          },
          label: removeCurrent
        } : null, /* Remove All */
        {
          key: 'removeAll',
          onClick: () => {
            onItemRemove === null || onItemRemove === void 0 ? void 0 : onItemRemove(getEnabledItemKeys(filteredItems));
          },
          label: removeAll
        }].filter(item => item);
      } else {
        items = [{
          key: 'selectAll',
          onClick: () => {
            const keys = getEnabledItemKeys(filteredItems);
            onItemSelectAll(keys, keys.length !== checkedKeys.length);
          },
          label: selectAll
        }, pagination ? {
          key: 'selectCurrent',
          onClick: () => {
            var _a;
            const pageItems = ((_a = this.defaultListBodyRef.current) === null || _a === void 0 ? void 0 : _a.getItems()) || [];
            onItemSelectAll(getEnabledItemKeys(pageItems.map(entity => entity.item)), true);
          },
          label: selectCurrent
        } : null, {
          key: 'selectInvert',
          onClick: () => {
            var _a;
            let availableKeys;
            if (pagination) {
              availableKeys = getEnabledItemKeys((((_a = this.defaultListBodyRef.current) === null || _a === void 0 ? void 0 : _a.getItems()) || []).map(entity => entity.item));
            } else {
              availableKeys = getEnabledItemKeys(filteredItems);
            }
            const checkedKeySet = new Set(checkedKeys);
            const newCheckedKeys = [];
            const newUnCheckedKeys = [];
            availableKeys.forEach(key => {
              if (checkedKeySet.has(key)) {
                newUnCheckedKeys.push(key);
              } else {
                newCheckedKeys.push(key);
              }
            });
            onItemSelectAll(newCheckedKeys, true);
            onItemSelectAll(newUnCheckedKeys, false);
          },
          label: selectInvert
        }];
      }
      const dropdown = /*#__PURE__*/React.createElement(Dropdown, {
        className: `${prefixCls}-header-dropdown`,
        menu: {
          items
        },
        disabled: disabled
      }, /*#__PURE__*/React.createElement(DownOutlined, null));
      // ================================== Render ===================================
      return /*#__PURE__*/React.createElement("div", {
        className: listCls,
        style: style
      }, /*#__PURE__*/React.createElement("div", {
        className: `${prefixCls}-header`
      }, showSelectAll ? /*#__PURE__*/React.createElement(React.Fragment, null, checkAllCheckbox, dropdown) : null, /*#__PURE__*/React.createElement("span", {
        className: `${prefixCls}-header-selected`
      }, this.getSelectAllLabel(checkedKeys.length, filteredItems.length)), /*#__PURE__*/React.createElement("span", {
        className: `${prefixCls}-header-title`
      }, titleText)), listBody, listFooter);
    }
  }]);
  return TransferList;
}(React.PureComponent);
export { TransferList as default };
