// antd/lib/transfer/ListBody.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OmitProps = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _createSuper2 = _interopRequireDefault(require("@babel/runtime/helpers/createSuper"));
var _classnames = _interopRequireDefault(require("classnames"));
var React = _interopRequireWildcard(require("react"));
var _pagination = _interopRequireDefault(require("../pagination"));
var _ListItem = _interopRequireDefault(require("./ListItem"));
const OmitProps = ['handleFilter', 'handleClear', 'checkedKeys'];
exports.OmitProps = OmitProps;
function parsePagination(pagination) {
  if (!pagination) {
    return null;
  }
  const defaultPagination = {
    pageSize: 10,
    simple: true,
    showSizeChanger: false,
    showLessItems: false
  };
  if (typeof pagination === 'object') {
    return Object.assign(Object.assign({}, defaultPagination), pagination);
  }
  return defaultPagination;
}
let ListBody = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(ListBody, _React$Component);
  var _super = (0, _createSuper2.default)(ListBody);
  function ListBody() {
    var _this;
    (0, _classCallCheck2.default)(this, ListBody);
    _this = _super.apply(this, arguments);
    _this.state = {
      current: 1
    };
    _this.onItemSelect = item => {
      const {
        onItemSelect,
        selectedKeys
      } = _this.props;
      const checked = selectedKeys.includes(item.key);
      onItemSelect(item.key, !checked);
    };
    _this.onItemRemove = item => {
      const {
        onItemRemove
      } = _this.props;
      onItemRemove === null || onItemRemove === void 0 ? void 0 : onItemRemove([item.key]);
    };
    _this.onPageChange = current => {
      _this.setState({
        current
      });
    };
    _this.getItems = () => {
      const {
        current
      } = _this.state;
      const {
        pagination,
        filteredRenderItems
      } = _this.props;
      const mergedPagination = parsePagination(pagination);
      let displayItems = filteredRenderItems;
      if (mergedPagination) {
        displayItems = filteredRenderItems.slice((current - 1) * mergedPagination.pageSize, current * mergedPagination.pageSize);
      }
      return displayItems;
    };
    return _this;
  }
  (0, _createClass2.default)(ListBody, [{
    key: "render",
    value: function render() {
      const {
        current
      } = this.state;
      const {
        prefixCls,
        onScroll,
        filteredRenderItems,
        selectedKeys,
        disabled: globalDisabled,
        showRemove,
        pagination
      } = this.props;
      const mergedPagination = parsePagination(pagination);
      let paginationNode = null;
      if (mergedPagination) {
        paginationNode = /*#__PURE__*/React.createElement(_pagination.default, {
          simple: mergedPagination.simple,
          showSizeChanger: mergedPagination.showSizeChanger,
          showLessItems: mergedPagination.showLessItems,
          size: "small",
          disabled: globalDisabled,
          className: `${prefixCls}-pagination`,
          total: filteredRenderItems.length,
          pageSize: mergedPagination.pageSize,
          current: current,
          onChange: this.onPageChange
        });
      }
      return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("ul", {
        className: (0, _classnames.default)(`${prefixCls}-content`, {
          [`${prefixCls}-content-show-remove`]: showRemove
        }),
        onScroll: onScroll
      }, this.getItems().map(_ref => {
        let {
          renderedEl,
          renderedText,
          item
        } = _ref;
        const {
          disabled
        } = item;
        const checked = selectedKeys.includes(item.key);
        return /*#__PURE__*/React.createElement(_ListItem.default, {
          disabled: globalDisabled || disabled,
          key: item.key,
          item: item,
          renderedText: renderedText,
          renderedEl: renderedEl,
          checked: checked,
          prefixCls: prefixCls,
          onClick: this.onItemSelect,
          onRemove: this.onItemRemove,
          showRemove: showRemove
        });
      })), paginationNode);
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(_ref2, _ref3) {
      let {
        filteredRenderItems,
        pagination
      } = _ref2;
      let {
        current
      } = _ref3;
      const mergedPagination = parsePagination(pagination);
      if (mergedPagination) {
        // Calculate the page number
        const maxPageCount = Math.ceil(filteredRenderItems.length / mergedPagination.pageSize);
        if (current > maxPageCount) {
          return {
            current: maxPageCount
          };
        }
      }
      return null;
    }
  }]);
  return ListBody;
}(React.Component);
var _default = ListBody;
exports.default = _default;
