// @arco-design/web-react/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Space = exports.Slider = exports.Skeleton = exports.Select = exports.Result = exports.ResizeBox = exports.Rate = exports.Radio = exports.Progress = exports.Portal = exports.Popover = exports.Popconfirm = exports.Pagination = exports.PageHeader = exports.Notification = exports.Modal = exports.Menu = exports.Message = exports.List = exports.Link = exports.Layout = exports.InputNumber = exports.InputTag = exports.Input = exports.Icon = exports.Grid = exports.Form = exports.Empty = exports.Dropdown = exports.Drawer = exports.Divider = exports.Descriptions = exports.DatePicker = exports.ConfigProvider = exports.Comment = exports.Collapse = exports.Checkbox = exports.Cascader = exports.Carousel = exports.Card = exports.Calendar = exports.Button = exports.Breadcrumb = exports.Badge = exports.BackTop = exports.Avatar = exports.AutoComplete = exports.Affix = exports.Anchor = exports.Alert = void 0;
exports.version = exports.Image = exports.Mentions = exports.Upload = exports.Typography = exports.Trigger = exports.TreeSelect = exports.Tree = exports.Transfer = exports.Tooltip = exports.TimePicker = exports.Timeline = exports.Tag = exports.Tabs = exports.Table = exports.Switch = exports.Steps = exports.Statistic = exports.Spin = void 0;
var Alert_1 = require("./Alert");
Object.defineProperty(exports, "Alert", { enumerable: true, get: function () { return __importDefault(Alert_1).default; } });
var Anchor_1 = require("./Anchor");
Object.defineProperty(exports, "Anchor", { enumerable: true, get: function () { return __importDefault(Anchor_1).default; } });
var Affix_1 = require("./Affix");
Object.defineProperty(exports, "Affix", { enumerable: true, get: function () { return __importDefault(Affix_1).default; } });
var AutoComplete_1 = require("./AutoComplete");
Object.defineProperty(exports, "AutoComplete", { enumerable: true, get: function () { return __importDefault(AutoComplete_1).default; } });
var Avatar_1 = require("./Avatar");
Object.defineProperty(exports, "Avatar", { enumerable: true, get: function () { return __importDefault(Avatar_1).default; } });
var BackTop_1 = require("./BackTop");
Object.defineProperty(exports, "BackTop", { enumerable: true, get: function () { return __importDefault(BackTop_1).default; } });
var Badge_1 = require("./Badge");
Object.defineProperty(exports, "Badge", { enumerable: true, get: function () { return __importDefault(Badge_1).default; } });
var Breadcrumb_1 = require("./Breadcrumb");
Object.defineProperty(exports, "Breadcrumb", { enumerable: true, get: function () { return __importDefault(Breadcrumb_1).default; } });
var Button_1 = require("./Button");
Object.defineProperty(exports, "Button", { enumerable: true, get: function () { return __importDefault(Button_1).default; } });
var Calendar_1 = require("./Calendar");
Object.defineProperty(exports, "Calendar", { enumerable: true, get: function () { return __importDefault(Calendar_1).default; } });
var Card_1 = require("./Card");
Object.defineProperty(exports, "Card", { enumerable: true, get: function () { return __importDefault(Card_1).default; } });
var Carousel_1 = require("./Carousel");
Object.defineProperty(exports, "Carousel", { enumerable: true, get: function () { return __importDefault(Carousel_1).default; } });
var Cascader_1 = require("./Cascader");
Object.defineProperty(exports, "Cascader", { enumerable: true, get: function () { return __importDefault(Cascader_1).default; } });
var Checkbox_1 = require("./Checkbox");
Object.defineProperty(exports, "Checkbox", { enumerable: true, get: function () { return __importDefault(Checkbox_1).default; } });
var Collapse_1 = require("./Collapse");
Object.defineProperty(exports, "Collapse", { enumerable: true, get: function () { return __importDefault(Collapse_1).default; } });
var Comment_1 = require("./Comment");
Object.defineProperty(exports, "Comment", { enumerable: true, get: function () { return __importDefault(Comment_1).default; } });
var ConfigProvider_1 = require("./ConfigProvider");
Object.defineProperty(exports, "ConfigProvider", { enumerable: true, get: function () { return __importDefault(ConfigProvider_1).default; } });
var DatePicker_1 = require("./DatePicker");
Object.defineProperty(exports, "DatePicker", { enumerable: true, get: function () { return __importDefault(DatePicker_1).default; } });
var Descriptions_1 = require("./Descriptions");
Object.defineProperty(exports, "Descriptions", { enumerable: true, get: function () { return __importDefault(Descriptions_1).default; } });
var Divider_1 = require("./Divider");
Object.defineProperty(exports, "Divider", { enumerable: true, get: function () { return __importDefault(Divider_1).default; } });
var Drawer_1 = require("./Drawer");
Object.defineProperty(exports, "Drawer", { enumerable: true, get: function () { return __importDefault(Drawer_1).default; } });
var Dropdown_1 = require("./Dropdown");
Object.defineProperty(exports, "Dropdown", { enumerable: true, get: function () { return __importDefault(Dropdown_1).default; } });
var Empty_1 = require("./Empty");
Object.defineProperty(exports, "Empty", { enumerable: true, get: function () { return __importDefault(Empty_1).default; } });
var Form_1 = require("./Form");
Object.defineProperty(exports, "Form", { enumerable: true, get: function () { return __importDefault(Form_1).default; } });
var Grid_1 = require("./Grid");
Object.defineProperty(exports, "Grid", { enumerable: true, get: function () { return __importDefault(Grid_1).default; } });
var Icon_1 = require("./Icon");
Object.defineProperty(exports, "Icon", { enumerable: true, get: function () { return __importDefault(Icon_1).default; } });
var Input_1 = require("./Input");
Object.defineProperty(exports, "Input", { enumerable: true, get: function () { return __importDefault(Input_1).default; } });
var InputTag_1 = require("./InputTag");
Object.defineProperty(exports, "InputTag", { enumerable: true, get: function () { return __importDefault(InputTag_1).default; } });
var InputNumber_1 = require("./InputNumber");
Object.defineProperty(exports, "InputNumber", { enumerable: true, get: function () { return __importDefault(InputNumber_1).default; } });
var Layout_1 = require("./Layout");
Object.defineProperty(exports, "Layout", { enumerable: true, get: function () { return __importDefault(Layout_1).default; } });
var Link_1 = require("./Link");
Object.defineProperty(exports, "Link", { enumerable: true, get: function () { return __importDefault(Link_1).default; } });
var List_1 = require("./List");
Object.defineProperty(exports, "List", { enumerable: true, get: function () { return __importDefault(List_1).default; } });
var Message_1 = require("./Message");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return __importDefault(Message_1).default; } });
var Menu_1 = require("./Menu");
Object.defineProperty(exports, "Menu", { enumerable: true, get: function () { return __importDefault(Menu_1).default; } });
var Modal_1 = require("./Modal");
Object.defineProperty(exports, "Modal", { enumerable: true, get: function () { return __importDefault(Modal_1).default; } });
var Notification_1 = require("./Notification");
Object.defineProperty(exports, "Notification", { enumerable: true, get: function () { return __importDefault(Notification_1).default; } });
var PageHeader_1 = require("./PageHeader");
Object.defineProperty(exports, "PageHeader", { enumerable: true, get: function () { return __importDefault(PageHeader_1).default; } });
var Pagination_1 = require("./Pagination");
Object.defineProperty(exports, "Pagination", { enumerable: true, get: function () { return __importDefault(Pagination_1).default; } });
var Popconfirm_1 = require("./Popconfirm");
Object.defineProperty(exports, "Popconfirm", { enumerable: true, get: function () { return __importDefault(Popconfirm_1).default; } });
var Popover_1 = require("./Popover");
Object.defineProperty(exports, "Popover", { enumerable: true, get: function () { return __importDefault(Popover_1).default; } });
var Portal_1 = require("./Portal");
Object.defineProperty(exports, "Portal", { enumerable: true, get: function () { return __importDefault(Portal_1).default; } });
var Progress_1 = require("./Progress");
Object.defineProperty(exports, "Progress", { enumerable: true, get: function () { return __importDefault(Progress_1).default; } });
var Radio_1 = require("./Radio");
Object.defineProperty(exports, "Radio", { enumerable: true, get: function () { return __importDefault(Radio_1).default; } });
var Rate_1 = require("./Rate");
Object.defineProperty(exports, "Rate", { enumerable: true, get: function () { return __importDefault(Rate_1).default; } });
var ResizeBox_1 = require("./ResizeBox");
Object.defineProperty(exports, "ResizeBox", { enumerable: true, get: function () { return __importDefault(ResizeBox_1).default; } });
var Result_1 = require("./Result");
Object.defineProperty(exports, "Result", { enumerable: true, get: function () { return __importDefault(Result_1).default; } });
var Select_1 = require("./Select");
Object.defineProperty(exports, "Select", { enumerable: true, get: function () { return __importDefault(Select_1).default; } });
var Skeleton_1 = require("./Skeleton");
Object.defineProperty(exports, "Skeleton", { enumerable: true, get: function () { return __importDefault(Skeleton_1).default; } });
var Slider_1 = require("./Slider");
Object.defineProperty(exports, "Slider", { enumerable: true, get: function () { return __importDefault(Slider_1).default; } });
var Space_1 = require("./Space");
Object.defineProperty(exports, "Space", { enumerable: true, get: function () { return __importDefault(Space_1).default; } });
var Spin_1 = require("./Spin");
Object.defineProperty(exports, "Spin", { enumerable: true, get: function () { return __importDefault(Spin_1).default; } });
var Statistic_1 = require("./Statistic");
Object.defineProperty(exports, "Statistic", { enumerable: true, get: function () { return __importDefault(Statistic_1).default; } });
var Steps_1 = require("./Steps");
Object.defineProperty(exports, "Steps", { enumerable: true, get: function () { return __importDefault(Steps_1).default; } });
var Switch_1 = require("./Switch");
Object.defineProperty(exports, "Switch", { enumerable: true, get: function () { return __importDefault(Switch_1).default; } });
var Table_1 = require("./Table");
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return __importDefault(Table_1).default; } });
var Tabs_1 = require("./Tabs");
Object.defineProperty(exports, "Tabs", { enumerable: true, get: function () { return __importDefault(Tabs_1).default; } });
var Tag_1 = require("./Tag");
Object.defineProperty(exports, "Tag", { enumerable: true, get: function () { return __importDefault(Tag_1).default; } });
var Timeline_1 = require("./Timeline");
Object.defineProperty(exports, "Timeline", { enumerable: true, get: function () { return __importDefault(Timeline_1).default; } });
var TimePicker_1 = require("./TimePicker");
Object.defineProperty(exports, "TimePicker", { enumerable: true, get: function () { return __importDefault(TimePicker_1).default; } });
var Tooltip_1 = require("./Tooltip");
Object.defineProperty(exports, "Tooltip", { enumerable: true, get: function () { return __importDefault(Tooltip_1).default; } });
var Transfer_1 = require("./Transfer");
Object.defineProperty(exports, "Transfer", { enumerable: true, get: function () { return __importDefault(Transfer_1).default; } });
var Tree_1 = require("./Tree");
Object.defineProperty(exports, "Tree", { enumerable: true, get: function () { return __importDefault(Tree_1).default; } });
var TreeSelect_1 = require("./TreeSelect");
Object.defineProperty(exports, "TreeSelect", { enumerable: true, get: function () { return __importDefault(TreeSelect_1).default; } });
var Trigger_1 = require("./Trigger");
Object.defineProperty(exports, "Trigger", { enumerable: true, get: function () { return __importDefault(Trigger_1).default; } });
var Typography_1 = require("./Typography");
Object.defineProperty(exports, "Typography", { enumerable: true, get: function () { return __importDefault(Typography_1).default; } });
var Upload_1 = require("./Upload");
Object.defineProperty(exports, "Upload", { enumerable: true, get: function () { return __importDefault(Upload_1).default; } });
var Mentions_1 = require("./Mentions");
Object.defineProperty(exports, "Mentions", { enumerable: true, get: function () { return __importDefault(Mentions_1).default; } });
var Image_1 = require("./Image");
Object.defineProperty(exports, "Image", { enumerable: true, get: function () { return __importDefault(Image_1).default; } });
exports.version = '2.44.0';
